holidayextras/generator-webapp-feature
# TAR-196 - building a generator for webapp features

#### What does this PR do?
This project is a [Yeoman generator](yeoman.io) which will builds features for our [webapp](https://github.com/holidayextras/webapp-template) projects.

For more info on this see the Readme within this PR.

Basically it looks for a package.json file and assumes that is your project root. It then expects a code directory and the code directory to contain a package.json file. If these exist then it will prompt you for the feature name. It then creates the necessary directory structure, writes a basic controller and updates the *code/index.js* file to tie in the newly created feature.

#### What unit or integration tests does this PR have?
I've written unit tests for this which are included within this PR.

#### What selenium tests does this PR have?
None

#### How should a developer review this?
- Check through the code and make sure you can easily follow it and that what it's doing makes sense to you.
- run the tests with the `npm test` command.

#### How should this be manually tested?
- install yeoman `npm i -g yo`
- install this generator using npm from repo or use `npm link`
- clone the webapp-template repo `git clone git@github.com:holidayextras/webapp-template.git`
- cd into the webapp-template dir `cd webapp-template`
- edit `code/index.js` adding the following:
    - `//====== yeoman requireHook =====//` below the require call for the productController
    - `//===== yeoman controllerHook =====//` within the `this.controller` declaration in `initialize()`
- run the generator `yo` (and follow instructions)  -or- `yo webapp-template`

check that the file structure defined in the README is created:

```
path/to/code
├ [featureName]
│ ├ controllers
│ │ └ [featureName]Controller.js
│ ├ style
│ ├ view_controllers
│ └ views
```

check that the `code/index.js` file is updated with references to the new feature.

If you modify the created controller so that it has an actual route defined and then run `make init && make dev` in the webapp-template project you should be able to navigate to that route and see the controller in action. For instance change 

```
routes: {
  "": "index"
},
```

to 
```
routes: {
  "test": "index"
},
```

and then in your browser navigate to *localhost:5000/index.html#test* and you should see the action of the controller in the browser (default is to add a `<div class="[featureName]"></div>` to the document.body).

#### Any background context you want to provide?
If not already familiar you could read docs on:
- [Yeoman](yeoman.io)
- [webapp-template](https://github.com/holidayextras/webapp-template) (Read comments in files as well)

#### What are the relevant tickets?
[TAR-196](https://holidayextras.jira.com/browse/TAR-196)

#### Screenshots (if appropriate)
None

#### What gif best describes how you feel about this work?
![](http://media.giphy.com/media/VUYTIed7MnjoY/giphy.gif)
---
#### Quality checklist (for PR author to complete BEFORE code review):

#### IP or Developer Review:

#### Software Engineer or Developer review:

#### Software Engineer or project guru review:
